Add more smart pointers to RenderSearchField and SearchPopup

abigailfox opened the pull request:
#### 2fa09212b231cb8c79895fcde70eebf92835601d
<pre>
Add more smart pointers to RenderSearchField and SearchPopup
<a href="https://bugs.webkit.org/show_bug.cgi?id=271504">https://bugs.webkit.org/show_bug.cgi?id=271504</a>
<a href="https://rdar.apple.com/125268281">rdar://125268281</a>

Reviewed by Ryosuke Niwa and Chris Dumez.

Add more smart pointers to RenderSearchField and modify supporting files and functions as needed

* Source/WebCore/loader/EmptyClients.cpp:
* Source/WebCore/platform/SearchPopupMenu.h:
* Source/WebCore/rendering/RenderSearchField.cpp:
(WebCore::RenderSearchField::willBeDestroyed):
(WebCore::RenderSearchField::resultsButtonElement const):
(WebCore::RenderSearchField::cancelButtonElement const):
(WebCore::RenderSearchField::addSearchResult):
(WebCore::RenderSearchField::showPopup):
(WebCore::RenderSearchField::hidePopup):
(WebCore::RenderSearchField::computeControlLogicalHeight const):
(WebCore::RenderSearchField::updateFromElement):
(WebCore::RenderSearchField::autosaveName const):
(WebCore::RenderSearchField::valueChanged):
(WebCore::RenderSearchField::clientPaddingLeft const):
(WebCore::RenderSearchField::clientPaddingRight const):
(WebCore::RenderSearchField::setTextFromItem):
(WebCore::RenderSearchField::createScrollbar):
* Source/WebKit/WebProcess/WebCoreSupport/WebSearchPopupMenu.cpp:
(WebKit::WebSearchPopupMenu::protectedPopupMenu):
* Source/WebKit/WebProcess/WebCoreSupport/WebSearchPopupMenu.h:
* Source/WebKitLegacy/ios/WebCoreSupport/SearchPopupMenuIOS.cpp:
(SearchPopupMenuIOS::protectedPopupMenu):
* Source/WebKitLegacy/ios/WebCoreSupport/SearchPopupMenuIOS.h:
* Source/WebKitLegacy/mac/WebCoreSupport/SearchPopupMenuMac.h:
* Source/WebKitLegacy/mac/WebCoreSupport/SearchPopupMenuMac.mm:
(SearchPopupMenuMac::protectedPopupMenu):

Canonical link: <a href="https://commits.webkit.org/276849@main">https://commits.webkit.org/276849@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/faa838a71003dc988f11703bc110002480bf5e34

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45760 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24887 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48341 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48430 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41795 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/48066 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29169 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/22285 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37467 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46338 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/21967 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39475 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18624 "Found 1 new API test failure: /WPE/TestResources:/webkit/WebKitWebView/sync-request-on-max-conns (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/19367 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/40566 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3804 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/42071 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40908 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50185 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/20751 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/17263 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/44583 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/22056 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43443 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10183 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/22415 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21745 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/loader/EmptyClients.cpp:
```diff
@@ -471,6 +471,7 @@ class EmptySearchPopupMenu : public SearchPopupMenu {
 
 private:
     PopupMenu* popupMenu() final { return m_popup.ptr(); }
+    RefPtr<PopupMenu> protectedPopupMenu() final { return m_popup.ptr(); }
```
FYI, using `m_popup.copyRef()` would be slightly more efficient. It avoids the RefPtr constructor having the null check the pointer, since we know it is non null.

But also, we don't need this override function at all.

cdumez commented on Source/WebCore/platform/SearchPopupMenu.h:
```diff
@@ -40,6 +40,7 @@ class SearchPopupMenu : public RefCounted<SearchPopupMenu> {
 public:
     virtual ~SearchPopupMenu() = default;
     virtual PopupMenu* popupMenu() = 0;
+    virtual RefPtr<PopupMenu> protectedPopupMenu() = 0;
```
This doesn't need to be virtual, you can do:
```
RefPtr<PopupMenu> protectedPopupMenu() { return popupMenu(); }
```

cdumez commented on Source/WebCore/rendering/RenderSearchField.cpp:
```diff
@@ -82,20 +82,23 @@ void RenderSearchField::willBeDestroyed()
 
 inline HTMLElement* RenderSearchField::resultsButtonElement() const
 {
-    return inputElement().resultsButtonElement();
```
We should introduce a `protectedInputElement()` and use it here, to avoid creating a new local variable.

cdumez commented on Source/WebCore/rendering/RenderSearchField.cpp:
```diff
@@ -82,20 +82,23 @@ void RenderSearchField::willBeDestroyed()
 inline HTMLElement* RenderSearchField::cancelButtonElement() const
 {
-    return inputElement().cancelButtonElement();
+    Ref input = inputElement();
```
ditto

cdumez commented on Source/WebCore/rendering/RenderSearchField.cpp:
```diff
@@ -146,24 +149,25 @@ void RenderSearchField::showPopup()
     IntRect absBounds = absoluteBoundingBoxRectIgnoringTransforms();
     absBounds.setLocation(roundedIntPoint(absTopLeft));
-    m_searchPopup->popupMenu()->show(absBounds, &view().frameView(), -1);
+    RefPtr popupMenu = m_searchPopup->protectedPopupMenu();
```
We're failing to protect m_searchPopup here, this should be something like:
```
protectedSearchPopup()->protectedPopupMenu()->show(absBounds, &view().frameView(), -1);
```

cdumez commented on Source/WebCore/rendering/RenderSearchField.cpp:
```diff
@@ -191,8 +195,10 @@ void RenderSearchField::updateFromElement()
-        m_searchPopup->popupMenu()->updateFromElement();
+    if (m_searchPopupIsVisible) {
+        RefPtr popupMenu = m_searchPopup->popupMenu();
+        popupMenu->updateFromElement();
```
`protectedSearchPopup()->protectedPopupMenu()->updateFromElement();`. We shouldn't need a new local variable.

cdumez commented on Source/WebCore/rendering/RenderSearchField.cpp:
```diff
@@ -218,7 +224,8 @@ Visibility RenderSearchField::visibilityForCancelButton() const
 {
-    return inputElement().attributeWithoutSynchronization(nameAttr);
+    Ref input = inputElement();
+    return input->attributeWithoutSynchronization(nameAttr);
```
`return protectedInputElement()->attributeWithoutSynchronization(nameAttr);`

We shouldn't need a new local variable.

rniwa commented on Source/WebCore/rendering/RenderSearchField.cpp:
```diff
@@ -82,20 +82,23 @@ void RenderSearchField::willBeDestroyed()
 inline HTMLElement* RenderSearchField::resultsButtonElement() const
 {
-    return inputElement().resultsButtonElement();
+    Ref input = inputElement();
```
Maybe we should add protectedInputElement()?

cdumez commented on Source/WebCore/rendering/RenderSearchField.cpp:
```diff
@@ -310,16 +318,19 @@ int RenderSearchField::clientInsetRight() const
-        if (RenderBox* innerBlockBox = innerBlockElement() ? innerBlockElement()->renderBox() : 0)
+    RefPtr container = containerElement();
+    RefPtr innerBlock = innerBlockElement();
+    if (RenderBox* containerBox = container ? container->renderBox() : 0) {
```
Please use `nullptr` instead of `0` since you're touching this line.

cdumez replied:
Also using `auto*` would be good.

cdumez commented on Source/WebCore/rendering/RenderSearchField.cpp:
```diff
@@ -310,16 +318,19 @@ int RenderSearchField::clientInsetRight() const
+    RefPtr container = containerElement();
+    RefPtr innerBlock = innerBlockElement();
+    if (RenderBox* containerBox = container ? container->renderBox() : 0) {
+        if (RenderBox* innerBlockBox = innerBlock ? innerBlock->renderBox() : 0)
```
ditto.

cdumez commented on Source/WebCore/rendering/RenderSearchField.cpp:
```diff
@@ -362,7 +373,8 @@ bool RenderSearchField::itemIsSelected(unsigned) const
 {
-    inputElement().setValue(itemText(listIndex));
+    Ref input = inputElement();
+    input->setValue(itemText(listIndex));
```
`protectedInputElement()->setValue(itemText(listIndex));`

rniwa commented on Source/WebCore/rendering/RenderSearchField.cpp:
```diff
@@ -378,8 +390,9 @@ HostWindow* RenderSearchField::hostWindow() const
 Ref<Scrollbar> RenderSearchField::createScrollbar(ScrollableArea& scrollableArea, ScrollbarOrientation orientation, ScrollbarWidth widthStyle)
 {
     bool usesLegacyScrollbarStyle = style().usesLegacyScrollbarStyle();
+    RefPtr input = &inputElement();
```
Why not Ref?

abigailfox replied:
I can do 
```    
Ref input = inputElement(); 
if (usesLegacyScrollbarStyle)
        return RenderScrollbar::createCustomScrollbar(scrollableArea, orientation, input.ptr());
```
if that's preferred. `createCustomScrollbar` needed a pointer

cdumez replied:
Yes, that's definitely preferred. No reason to use a nullable pointer type (RefPtr) when you can use a non-nullable one (Ref).

rniwa commented on Source/WebCore/rendering/RenderSearchField.cpp:
```diff
@@ -378,8 +390,9 @@ HostWindow* RenderSearchField::hostWindow() const
+    RefPtr input = &inputElement();
     if (usesLegacyScrollbarStyle)
-        return RenderScrollbar::createCustomScrollbar(scrollableArea, orientation, &inputElement());
+        return RenderScrollbar::createCustomScrollbar(scrollableArea, orientation, input.get());
```
You can call input.ptr() once you've changed the above to use Ref instead.

cdumez replied:
`protectedInputElement().ptr()`

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebSearchPopupMenu.cpp:
```diff
@@ -47,6 +47,11 @@ PopupMenu* WebSearchPopupMenu::popupMenu()
 
+RefPtr<PopupMenu> WebSearchPopupMenu::protectedPopupMenu()
+{
+    return m_popup.get();
```
`.get()` is not needed. Also, the function can probably be inlined in the header given that it is so trivial.

cdumez commented on Source/WebKitLegacy/ios/WebCoreSupport/SearchPopupMenuIOS.cpp:
```diff
@@ -35,6 +35,11 @@ PopupMenu* SearchPopupMenuIOS::popupMenu()
 
+RefPtr<PopupMenu> SearchPopupMenuIOS::protectedPopupMenu()
+{
+    return m_popup.ptr();
```
ditto.

cdumez commented on Source/WebKitLegacy/mac/WebCoreSupport/SearchPopupMenuMac.mm:
```diff
@@ -49,6 +49,11 @@ static String defaultSearchFieldRecentSearchesStorageDirectory()
 
+RefPtr<WebCore::PopupMenu> SearchPopupMenuMac::protectedPopupMenu()
+{
+    return m_popup.get();
```
ditto.

cdumez commented on Source/WebCore/rendering/RenderSearchField.h:
```diff
@@ -79,6 +79,7 @@ class RenderSearchField final : public RenderTextControlSingleLine, private Popu
     FontSelector* fontSelector() const override;
     HostWindow* hostWindow() const override;
     Ref<Scrollbar> createScrollbar(ScrollableArea&, ScrollbarOrientation, ScrollbarWidth) override;
+    RefPtr<SearchPopupMenu> protectedSearchPopup() { return m_searchPopup; };
```
Function should probably be `const`.

cdumez commented on Source/WebKitLegacy/mac/WebCoreSupport/SearchPopupMenuMac.h:
```diff
@@ -35,6 +35,7 @@ class SearchPopupMenuMac : public WebCore::SearchPopupMenu {
     ~SearchPopupMenuMac();
 
     WebCore::PopupMenu* popupMenu() override;
+    RefPtr<WebCore::PopupMenu> protectedPopupMenu() { return popupMenu(); }
```
I don't think we need this?

cdumez commented on Source/WebKitLegacy/ios/WebCoreSupport/SearchPopupMenuIOS.h:
```diff
@@ -33,6 +33,7 @@ class SearchPopupMenuIOS : public WebCore::SearchPopupMenu {
     SearchPopupMenuIOS(WebCore::PopupMenuClient*);
 
     WebCore::PopupMenu* popupMenu() override;
+    RefPtr<WebCore::PopupMenu> protectedPopupMenu() { return popupMenu(); }
```
I don't think we need this?

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebSearchPopupMenu.h:
```diff
@@ -31,6 +31,7 @@ class WebSearchPopupMenu : public WebCore::SearchPopupMenu {
     static Ref<WebSearchPopupMenu> create(WebPage*, WebCore::PopupMenuClient*);
 
     WebCore::PopupMenu* popupMenu() override;
+    RefPtr<WebCore::PopupMenu> protectedPopupMenu() { return popupMenu(); }
```
I don't think we need this?

abigailfox replied:
Sorry, originally it was complaining about these, but I think it was a consequence of the way I wrote it previously. It seems fine without them now!

cdumez commented on Source/WebCore/rendering/RenderSearchField.cpp:
```diff
@@ -73,7 +73,7 @@ RenderSearchField::~RenderSearchField()
 void RenderSearchField::willBeDestroyed()
 {
     if (m_searchPopup) {
```
The way we usually write this is:
```
if (RefPtr searchPopup = std::exchange(m_searchPopup, nullptr))
    searchPopup->popupMenu()->disconnectClient();
```

cdumez commented on Source/WebCore/rendering/RenderSearchField.cpp:
```diff
@@ -82,20 +82,20 @@ void RenderSearchField::willBeDestroyed()
 void RenderSearchField::addSearchResult()
 {
-    if (inputElement().maxResults() <= 0)
+    if (protectedInputElement()->maxResults() <= 0)
```
You need a local variable, we don't want to call `protectedInputElement()` several times in the same scope, this is bad for perf.

```
Ref inputElement = this->inputElement();
// Then use inputElement in this function.
```

cdumez commented on Source/WebCore/rendering/RenderSearchField.cpp:
```diff
@@ -310,16 +311,19 @@ int RenderSearchField::clientInsetRight() const
     LayoutUnit padding = paddingLeft();
-    if (RenderBox* box = innerBlockElement() ? innerBlockElement()->renderBox() : 0)
+    RefPtr innerBlock = innerBlockElement();
+    if (RenderBox* box = innerBlock ? innerBlock->renderBox() : 0)
```
`0` -> `nullptr`

cdumez replied:
also `RenderBox*` -> `auto*`

cdumez commented on Source/WebCore/rendering/RenderSearchField.cpp:
```diff
@@ -310,16 +311,19 @@ int RenderSearchField::clientInsetRight() const
-    if (RenderBox* containerBox = containerElement() ? containerElement()->renderBox() : 0) {
-        if (RenderBox* innerBlockBox = innerBlockElement() ? innerBlockElement()->renderBox() : 0)
+    RefPtr container = containerElement();
+    RefPtr innerBlock = innerBlockElement();
```
This should be moved inside the `if` scope, no reason to ref this if we don't end up using it.

abigailfox replied:
~Don't we need to have them exist in order to execute the if statement?~ Never mind, I think I misunderstood

cdumez replied:
Is this clearer?
```
RefPtr container = containerElement();
if (auto* containerBox = container ? container->renderBox() : nullptr) {
    RefPtr innerBlock = innerBlockElement();
    if (auto* innerBlockBox = innerBlock ? innerBlock->renderBox() : nullptr)
```

cdumez commented on Source/WebCore/rendering/RenderSearchField.cpp:
```diff
@@ -108,10 +107,10 @@ void RenderSearchField::addSearchResult()
         m_recentSearches.removeLast();
 
-    const AtomString& name = autosaveName();
+    AtomString name = autosaveName();
```
Why this change?

cdumez reviewed (approved):
Looks good now, except for one minor AtomString change that seems unrelated to this patch and should probably be reverted.

abigailfox committed:
Add more smart pointers to RenderSearchField and SearchPopup
https://bugs.webkit.org/show_bug.cgi?id=271504
rdar://125268281

Reviewed by Ryosuke Niwa and Chris Dumez.

Add more smart pointers to RenderSearchField and modify supporting files and functions as needed

* Source/WebCore/loader/EmptyClients.cpp:
* Source/WebCore/platform/SearchPopupMenu.h:
* Source/WebCore/rendering/RenderSearchField.cpp:
(WebCore::RenderSearchField::willBeDestroyed):
(WebCore::RenderSearchField::resultsButtonElement const):
(WebCore::RenderSearchField::cancelButtonElement const):
(WebCore::RenderSearchField::addSearchResult):
(WebCore::RenderSearchField::showPopup):
(WebCore::RenderSearchField::hidePopup):
(WebCore::RenderSearchField::computeControlLogicalHeight const):
(WebCore::RenderSearchField::updateFromElement):
(WebCore::RenderSearchField::autosaveName const):
(WebCore::RenderSearchField::valueChanged):
(WebCore::RenderSearchField::clientPaddingLeft const):
(WebCore::RenderSearchField::clientPaddingRight const):
(WebCore::RenderSearchField::setTextFromItem):
(WebCore::RenderSearchField::createScrollbar):
* Source/WebKit/WebProcess/WebCoreSupport/WebSearchPopupMenu.cpp:
(WebKit::WebSearchPopupMenu::protectedPopupMenu):
* Source/WebKit/WebProcess/WebCoreSupport/WebSearchPopupMenu.h:
* Source/WebKitLegacy/ios/WebCoreSupport/SearchPopupMenuIOS.cpp:
(SearchPopupMenuIOS::protectedPopupMenu):
* Source/WebKitLegacy/ios/WebCoreSupport/SearchPopupMenuIOS.h:
* Source/WebKitLegacy/mac/WebCoreSupport/SearchPopupMenuMac.h:
* Source/WebKitLegacy/mac/WebCoreSupport/SearchPopupMenuMac.mm:
(SearchPopupMenuMac::protectedPopupMenu):

Canonical link: https://commits.webkit.org/276849@main
